controllers/capabilities/tenant_internal_reconciler.go

```diff
@@ -72,12 +83,48 @@ func (r *TenantInternalReconciler) reconcileTenant() (*porta_client_pkg.Tenant,
 		return nil, err
 	}
 
-	if tenantDef == nil {
+	// if tenant is found in 3scale and deletion timestamp is present, attempt to delete
+	if tenantDef != nil && r.tenantR.DeletionTimestamp != nil {
+		// only try deleting tenant if it's not already marked for deletion
+		isTenantDeleted, err := r.confirmTenantDeleted(tenantDef)
+		if err != nil {
+			return nil, err
+		}
+
+		if !isTenantDeleted {
+			r.portaClient.DeleteTenant(r.tenantR.Status.TenantId)
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		// re-fetch tenant to get current state
+		tenantDef, err = r.fetchTenant()
+		if err != nil {
+			return nil, err
+		}
+
+		isTenantDeleted, err = r.confirmTenantDeleted(tenantDef)
+		if err != nil {
+			return nil, err
+		}
+
+		// fail if tenant is not marked for deletion
+		if !isTenantDeleted {
+			return nil, fmt.Errorf("tenant deletion failed - re-trying")
+		}
+	}
+
+	// create tenant if tenant is NOT found in 3scale and deletion timestamp is not present
+	if tenantDef == nil && r.tenantR.Finalizers == nil {
```

Comment:
Should the second check here be for Deletion timestamp. If not I think the comment needs an update.